src/App.tsx

```diff
@@ -1,24 +1,235 @@
+    setProcessingTodoIds(prevTodoIds => [...prevTodoIds, newId]);
+  };
+
+  const handleAddTodo = (todoTitle: string) => {
+    setTempTodo({
+      id: 0,
+      title: todoTitle,
+      userId: 0,
+      completed: false,
+    });
+
+    return todoService
+      .addTodos(todoTitle)
+      .then((newTodo) => {
+        setTodos((prevTodos) => [...prevTodos, newTodo]);
+      })
+      .catch(() => {
+        setErrorMessage('Unable to add a todo');
+        throw new Error();
+      })
+      .finally(() => {
+        setTempTodo(null);
+      });
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setNewIds(todoId);
+
+    todoService
+      .deleteTodos(todoId)
+      .then((() => {
+        setTodos((prevTodos) => prevTodos.filter(todo => todo.id !== todoId));
+      }))
+      .catch(() => {
+        setErrorMessage('Unable to delete a todo');
+      })
+      .finally(() => {
+        filterTodosIds(todoId);
+      });
+  };
+
+  const handleRenameTodo = (todo: Todo, newTodoTitle: string) => {
+    setNewIds(todo.id);
+
+    todoService
+      .updateTodo({
+        ...todo,
+        title: newTodoTitle,
+      })
+      .then(updatedTodo => {
+        setTodos(prevState => prevState.map(currentTodo => {
+          return currentTodo.id !== updatedTodo.id
+            ? currentTodo
+            : updatedTodo;
+        }));
+      })
+      .catch(() => {
+        setErrorMessage('Unable to update a todo');
+      })
+      .finally(() => {
+        filterTodosIds(todo.id);
+      });
+  };
+
+  const handleChangeStatusTodo = (todo: Todo) => {
+    setNewIds(todo.id);
+
+    todoService
+      .updateTodo({
+        ...todo,
+        completed: !todo.completed,
+      })
+      .then(updatedTodo => {
+        setTodos(prevState => prevState.map(currentTodo => {
+          return currentTodo.id !== updatedTodo.id ? currentTodo : updatedTodo;
+        }));
+      })
+      .catch(() => {
+        setErrorMessage('Unable to update a todo');
+      })
+      .finally(() => {
+        filterTodosIds(todo.id);
+      });
+  };
+
+  const handleClearCompletedTodos = () => {
+    todos
+      .filter(todo => todo.completed)
+      .forEach(todo => {
+        handleDeleteTodo(todo.id);
+      });
+  };
+
+  const visibleTodos = useMemo(() => {
+    return getFilteredTodos(todos, status);
+  }, [todos, status]);
+
+  const handleFilterStatus = (todosFilter: Status) => (
+    setStatus(todosFilter));
+
+  const allTodosCount = todos.length;
+  const CompletedTodosCount = todos.filter(({ completed }) => completed).length;
+  const isAllTodosCompleted = allTodosCount === CompletedTodosCount;
+
+  const handleChangeStatusTodos = () => {
+    if (isAllTodosCompleted) {
+      todos
+        .forEach(todo => {
+          handleChangeStatusTodo(todo);
+        });
+    } else {
+      todos
+        .filter(({ completed }) => !completed)
+        .forEach(todo => {
+          handleChangeStatusTodo(todo);
+        });
+    }
```

Comment:
Let's use `Promise.all`. Also, you repeat your code, you don't need `if/else`. Better firstly to filter todos if needed and then just use `handleChangeStatusTodo` on every todo.